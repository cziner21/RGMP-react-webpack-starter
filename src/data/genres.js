const Genres = ['All', 'Documentary', 'Comedy', 'Horror', 'Crime']

export default Genres
