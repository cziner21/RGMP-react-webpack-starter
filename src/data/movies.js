import PropTypes from 'prop-types'
import Genres from './genres'

const MovieData = {}
MovieData.propTypes = {
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    releaseDate: PropTypes.number.isRequired,
    genres: PropTypes.arrayOf(PropTypes.string).isRequired,
}

const Movies = []
for (let i = 10; i < 30; i++) {
    Movies.push({
        id: (i + 1).toString(),
        title: `Movie ${i}`,
        releaseDate: getRandom(2020, 1950),
        genres: Genres.sort(() => Math.random() - Math.random()).slice(
            0,
            getRandom(4, 0)
        ),
    })
}

function getRandom(min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min
}

export { MovieData, Movies }
