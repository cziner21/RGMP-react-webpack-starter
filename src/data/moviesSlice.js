import { createSlice, createAsyncThunk } from '@reduxjs/toolkit'

import axios from 'axios'

export const MoviesStatuses = {
    idle: 'idle',
    loading: 'loading',
    succeeded: 'succeeded',
    failed: 'failed',
}

const BASE_URL = 'http://localhost:4000/movies'

export const defaultSearchParams = {
    sortBy: 'title',
    sortOrder: 'asc',
    search: null,
    searchBy: null,
    filter: null,
    offset: null,
    limit: 100,
}

const initialState = {
    data: [],
    genres: [],
    searchParams: defaultSearchParams,
    selectedGenres: [],
    selectedMovieId: null,
    status: MoviesStatuses.idle, // 'idle' | 'loading' | 'succeeded' | 'failed'
    error: null,
    currentMovie: {},
}

export const fetchMovies = createAsyncThunk(
    'movies/fetchMovies',
    async (searchParams) => {
        const params = {
            sortBy: searchParams.sortBy ? searchParams.sortBy : 'title',
            sortOrder: searchParams.sortOrder ? searchParams.sortOrder : 'asc',
            search: searchParams.search ? searchParams.search : null,
            searchBy: searchParams.searchBy ? searchParams.searchBy : null,
            filter: searchParams.filter ? searchParams.filter : null,
            offset: searchParams.offset ? searchParams.offset : null,
            limit: searchParams.limit ? searchParams.limit : 100,
        }

        try {
            const response = await axios.get(BASE_URL, {
                params: params,
            })
            return response.data
        } catch (error) {
            return error.message
        }
    }
)

export const createMovie = createAsyncThunk(
    'movies/createMovie',
    async (movie) => {
        try {
            const response = await axios.post(BASE_URL, movie)
            return response.data
        } catch (error) {
            return error.message
        }
    }
)

export const updateMovie = createAsyncThunk(
    'movies/updateMovie',
    async (movie) => {
        try {
            const response = await axios.put(BASE_URL, movie)
            return response.data
        } catch (error) {
            return error.message
        }
    }
)

export const deleteMovie = createAsyncThunk(
    'movies/deleteMovie',
    async (id) => {
        try {
            const response = await axios.delete(`${BASE_URL}/${id}`)
            return response.data
        } catch (error) {
            return error.message
        }
    }
)

const moviesSlice = createSlice({
    name: 'movies',
    initialState: initialState,
    reducers: {
        setSearchParams: (state, action) => {
            state.searchParams = action.payload
        },
        setSelectedGenres: (state, action) => {
            state.selectedGenres = action.payload
        },
        setSelectedMovieId: (state, action) => {
            state.selectedMovieId = action.payload
        },
    },
    extraReducers(builder) {
        builder
            .addCase(fetchMovies.pending, (state, action) => {
                state.status = MoviesStatuses.loading
            })
            .addCase(fetchMovies.fulfilled, (state, action) => {
                const { totalAmount, data } = action.payload
                state.status = MoviesStatuses.succeeded
                state.data = data

                const genres = data
                    .map((x) => x.genres)
                    .reduce((a, b) => {
                        return a.concat(b)
                    }, [])
                state.genres = Array.from(new Set(genres))
            })
            .addCase(fetchMovies.rejected, (state, action) => {
                state.status = MoviesStatuses.failed
                state.error = action.error.message
            })
            .addCase(updateMovie.fulfilled, (state, action) => {
                console.log(action.payload)
                // todo refresh list, give notification
            })
            .addCase(updateMovie.rejected, (state, action) => {
                state.error = action.error.message
            })
            .addCase(createMovie.fulfilled, (state, action) => {
                console.log(action.payload)
                // todo refresh list, give notification
            })
            .addCase(createMovie.rejected, (state, action) => {
                state.error = action.error.message
            })
            .addCase(deleteMovie.fulfilled, (state, action) => {
                console.log(action.payload)
                // todo refresh list, give notification
            })
            .addCase(deleteMovie.rejected, (state, action) => {
                state.error = action.error.message
            })
    },
})

export const getSelectedMovieId = (state) => state.movies.selectedMovieId
export const getSelectedGenres = (state) => state.movies.selectedGenres
export const getSearchParams = (state) => state.movies.searchParams
export const allMovies = (state) => state.movies.data
export const allGenres = (state) => state.movies.genres
export const currentMovie = (state) => state.movies.currentMovie
export const getMoviesStatus = (state) => state.movies.status
export const getMoviesError = (state) => state.movies.error

export const { setSearchParams, setSelectedGenres, setSelectedMovieId } =
    moviesSlice.actions

export default moviesSlice.reducer
