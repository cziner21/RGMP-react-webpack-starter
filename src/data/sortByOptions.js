export const sortByOptions = [
    {
        id: 'title',
        name: 'Name',
    },
    {
        id: 'releaseDate',
        name: 'Release date',
    },
]
