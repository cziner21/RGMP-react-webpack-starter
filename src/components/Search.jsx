import React from 'react'
import styled from 'styled-components'
import { useDispatch } from 'react-redux'

import { setSearchParams, defaultSearchParams } from '../data/moviesSlice'

import StyledInput from './Input.jsx'
import { Button } from './Button.jsx'

const SearchBar = styled.div`
    display: flex;

    & input {
        margin-right: 0.5em;
    }
`

function Search() {
    const dispatch = useDispatch()

    const handleButtonClicked = (value) => {
        console.log(`Search button clicked ${value}`)
        dispatch(
            setSearchParams({
                ...defaultSearchParams,
                search: value,
                searchBy: 'title',
            })
        )
    }

    return (
        <SearchBar>
            <StyledInput
                placeholder={'What do you want to watch?'}
                onEnterPressed={(value) => handleButtonClicked(value)}
            />
            <Button onClick={() => handleButtonClicked()}>Search</Button>
        </SearchBar>
    )
}

export default Search
