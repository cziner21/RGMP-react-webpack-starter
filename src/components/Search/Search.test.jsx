import React from 'react'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import { MemoryRouter } from 'react-router-dom'
import { fireEvent, render, screen } from '@testing-library/react'

import { defaultSearchParams, MoviesStatuses } from '../../data/moviesSlice'
import Search from './Search.jsx'

describe('Search panel interactive tests', () => {
    const initialState = {
        data: [],
        genres: [],
        searchQuery: defaultSearchParams,
        selectedGenres: [],
        selectedMovieId: null,
        isSingleMovie: false,
        status: MoviesStatuses.idle,
        error: null,
        currentMovie: {},
    }
    const mockStore = configureStore()
    let store

    it('Renders and match snapshot', () => {
        store = mockStore(initialState)

        const searchBar = renderer
            .create(
                <Provider store={store}>
                    <MemoryRouter>
                        <Search />
                    </MemoryRouter>
                </Provider>
            )
            .toJSON()
        expect(searchBar).toMatchSnapshot()
    })

    it('Updates on change', () => {
        store = mockStore(initialState)

        const { container } = render(
            <Provider store={store}>
                <MemoryRouter>
                    <Search />
                </MemoryRouter>
            </Provider>
        )

        const inputElement = container.querySelector(`input[name="search"]`)

        fireEvent.change(inputElement, { target: { value: 'Rambo' } })

        expect(inputElement.value).toBe('Rambo')
    })
})
