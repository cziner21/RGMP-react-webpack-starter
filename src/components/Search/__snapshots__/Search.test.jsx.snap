// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search panel interactive tests Renders and match snapshot 1`] = `
<div
  className="sc-hKMtZM jYQeGD"
>
  <form
    action="#"
    onReset={[Function]}
    onSubmit={[Function]}
    style={
      Object {
        "display": "contents",
      }
    }
  >
    <input
      name="search"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="What do you want to watch?"
      style={
        Object {
          "backgroundColor": "#424242",
          "border": "solid 1px #424242",
          "borderRadius": "3px",
          "color": "#555555",
          "outline": "none",
          "padding": "0.75em",
          "width": "100%",
        }
      }
      value=""
    />
  </form>
  <button
    className="sc-bczRLJ jIzewS"
    onClick={[Function]}
  >
    Search
  </button>
</div>
`;
